File: frontend/public/src/service/lang.js
import axios from "axios";

export const LangService = new class {
	getLangs() {
		return new Promise((resolve, reject) => {
			axios.get(`${EMVI_WIKI_BACKEND_HOST}/api/v1/lang`)
			.then(r => {
				resolve(r.data);
			})
			.catch(e => {
				reject(e);
			});
		});
	}

	getLang(id) {
		return new Promise((resolve, reject) => {
			axios.get(`${EMVI_WIKI_BACKEND_HOST}/api/v1/lang/${id}`)
			.then(r => {
				resolve(r.data);
			})
			.catch(e => {
				reject(e);
			});
		});
	}

	addLang(code) {
		return new Promise((resolve, reject) => {
			axios.post(`${EMVI_WIKI_BACKEND_HOST}/api/v1/lang`, {code})
			.then(r => {
				resolve(r.data);
			})
			.catch(e => {
				reject(e);
			});
		});
	}

	switchDefault(language_id) {
		return new Promise((resolve, reject) => {
			axios.put(`${EMVI_WIKI_BACKEND_HOST}/api/v1/lang`, {language_id})
			.then(r => {
				resolve(r.data);
			})
			.catch(e => {
				reject(e);
			});
		});
	}
};

File: backend/api/lang.go
package api

import (
	"emviwiki/backend/context"
	"emviwiki/backend/lang"
	"emviwiki/shared/rest"
	"github.com/emvi/hide"
	"net/http"
)

func GetLangsHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	resp := lang.GetLangs(ctx.Organization)
	rest.WriteResponse(w, resp)
	return nil
}

func GetLangHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	id, err := rest.IdParam(r, "id")

	if err != nil {
		return []error{err}
	}

	language, err := lang.GetLang(ctx.Organization, id)

	if err != nil {
		return []error{err}
	}

	rest.WriteResponse(w, language)
	return nil
}

func AddLangHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	req := struct {
		Code string `json:"code"`
	}{}

	if err := rest.DecodeJSON(r, &req); err != nil {
		return []error{err}
	}

	if err := lang.AddLang(ctx.Organization, ctx.UserId, req.Code); err != nil {
		return []error{err}
	}

	return nil
}

func SwitchDefaultLangHandler(ctx context.EmviContext, w http.ResponseWriter, r *http.Request) []error {
	req := struct {
		LanguageId hide.ID `json:"language_id"`
	}{}

	if err := rest.DecodeJSON(r, &req); err != nil {
		return []error{err}
	}

	if err := lang.SwitchDefaultLanguage(ctx.Organization, ctx.UserId, req.LanguageId); err != nil {
		return []error{err}
	}

	return nil
}
